### ISSUE TITLE
Overly aggressive "unterminated '#pragma clang attribute push' at end of file" error

### ISSUE THREAD

**Issue by tweenk** [llvm/llvm-project#63859]
This error should be reported only at the end of a translation unit, not at the end of a file. Otherwise it prevents the following useful usage pattern:

some_header.h:

```
#include "defs.inc"
inline uint32_t add_with_rollover(uint32_t x)  {
  return x + 1;
}
#include "undefs.inc"
```

defs.inc:

```
#pragma clang attribute push(__attribute__((no_sanitize("unsigned-integer-overflow)), apply_to=function)
```

undefs.inc:

```
#pragma clang attribute pop
```

Currently this will give an error at the end of defs.inc, but I think it should not give an error at all. It should only give an error if the pop is missing from undefs.inc.

**Comment by tweenk** [comment]
Another reason for only reporting this error at the translation unit level is that it would be consistent with other push/pop pragmas, such as `#pragma clang diagnostic push`.

**Comment by tweenk** [comment]
Some additional context on the real-world reason why I wanted to use this pattern.

The Protocol Buffers library uses unsigned module arithmetic fairly extensively, including in headers, since this is technically fully-defined behavior. However, some Android OEMs want to compile code that uses Protocol Buffers with unsigned integer sanitization enabled.

We already have files called port_def.inc and port_undef.inc that we use as shown above to modify the header environment, so I was trying to add a `#pragma clang attribute push` directive disabling this sanitizer to port_def.inc. This would have let me remove all of the Android-specific local patches to use __buitin_add_overflow in header code. However, this issue prevents the idea from working.

AOSP change which shows what I've been trying to do:
https://android-review.googlesource.com/c/platform/external/protobuf/+/2658056

**Comment by shafik** [comment]
CC @AaronBallman who may know why it works this way

**Comment by Endilll** [comment]
I wonder if it's related to his commit 7ff507f1448bfdfcaa91d177d1f655dcb17557e7

**Comment by tweenk** [comment]
Not sure if the linked commit is relevant, it's about `#pragma clang diagnostic`, which doesn't have the behavior this bug is about. In fact I already have a `#pragma GCC diagnostic push` in port_def.inc and it works as expected.

Based on the review history at https://reviews.llvm.org/D30009 it might have behaved like this from the start. Comment for diff 94208 says "The `unterminated '#pragma clang attribute push' at end of file` diagnostic is now an error."

**Comment by fowles** [comment]
This pattern is used pretty heavily in protobuf.  https://github.com/protocolbuffers/protobuf/blob/0ea00526a11c405b342463cf75f40ca5753d4a0c/src/google/protobuf/port_def.inc#L1047

**Comment by tweenk** [comment]
I was too quick to file this. The problem I was observing is actually caused by the fact that some Protobuf .cc files do not include port_undef.inc at the end, it only consistently includes it in headers.

I'll close this and address the issue in Protobuf upstream.
